src/core/sensor/qgsabstractsensor.h

```diff
@@ -0,0 +1,205 @@
+      QDateTime lastTimestamp;
+    };
+
+    /**
+     * Constructor for an abstract sensor, bound to the specified \a parent.
+     */
+    explicit QgsAbstractSensor( QObject *parent SIP_TRANSFERTHIS = nullptr );
+    virtual ~QgsAbstractSensor() = default;
+
+    /**
+     * Returns the sensor ID.
+     */
+    QString id() const { return mId; }
+
+    /**
+     * Returns the sensor type
+     */
+    virtual QString type() const { return QString(); }
+
+    /**
+     * Returns the sensor name
+     */
+    QString name() const;
+
+    /**
+     * Sets the sensor name
+     */
+    void setName( const QString &name );
+
+    /**
+     * Returns the current sensor status.
+     */
+    Qgis::SensorStatus status() const;
+
+    /**
+     * Sets the current sensor \a status.
+     */
+    void setStatus( Qgis::SensorStatus status );
+
+    /**
+     * Connects the sensor to its source.
+     */
+    void connectSensor();
+
+    /**
+     * Disconnects the sensor from its source.
+     */
+    void disconnectSensor();
+
+    /**
+     * Returns the latest captured date from the sensor.
+     */
+    QgsAbstractSensor::SensorData data() const;
+
+    /**
+     * Write generic sensor properties into a DOM element.
+     * \param parentElement parent DOM element (e.g 'Sensors' element)
+     * \param document DOM document
+     */
+    bool writeXml( QDomElement &parentElement, QDomDocument &document ) const;
+
+    /**
+     * Restores generic sensor details from a DOM element.
+     * \param element DOM node corresponding to item (e.g. 'Sensor' element)
+     * \param document DOM document
+     */
+    bool readXml( const QDomElement &element, const QDomDocument &document );
+
+    /**
+     * Write specific sensor type properties into a DOM element.
+     * \param element  DOM node corresponding to item (e.g. 'Sensor' element)
+     * \param document DOM document
+     */
+    virtual bool writePropertiesToElement( QDomElement &element, QDomDocument &document ) const;
+
+    /**
+     * Restores specific sensor type properties from a DOM element.
+     * \param element  DOM node corresponding to item (e.g. 'Sensor' element)
+     * \param document DOM document
+     */
+    virtual bool readPropertiesFromElement( const QDomElement &element, const QDomDocument &document );
+
+  signals:
+
+    //! Emitted when the sensor name has changed.
+    void nameChanged();
+
+    //! Emitted when the sensor status has changed.
+    void statusChanged();
+
+    //! Emitted when the captured sensor data has changed.
+    void dataChanged();
+
+  protected:
+
+    /**
+     * Handles the connection to the sensor.
+     * \note Triggered by calling connectSensor()
+     */
+    virtual void handleConnect() = 0;
+
+    /**
+     * Handles the disconnection from the sensor.
+     * \note Triggered by calling disconnectSensor()
+     */
+    virtual void handleDisconnect() = 0;
+
+    QgsAbstractSensor::SensorData mData;
```

Comment:
Is it possible from python to access a protected member? I thought there was a problem with SIP about that and it is only possible to access protected functions but maybe I am wrong.

But if it is the case there should be a function to set mData, otherwise it will not be possible to implement custom sensors in a plugin.